File: client/app/src/main/java/com/yszln/qiuqiu/ui/main/view/behavior/HeaderSearchBehavior.java
package com.yszln.qiuqiu.ui.main.view.behavior;

import android.content.Context;
import android.content.res.Resources;
import android.util.AttributeSet;
import android.view.View;


import androidx.coordinatorlayout.widget.CoordinatorLayout;

import com.yszln.qiuqiu.R;


/**
 * @author LiBinBin
 * @create 2020/7/29
 * @Describe
 */
public class HeaderSearchBehavior extends CoordinatorLayout.Behavior<View> {


    public HeaderSearchBehavior(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    public boolean layoutDependsOn(CoordinatorLayout parent, View child, View dependency) {
        if (dependency != null && dependency.getId() == R.id.image) {
            return true;
        }
        return false;
    }

    @Override
    public boolean onDependentViewChanged(CoordinatorLayout parent, View child, View dependency) {
        if (dependency != null && dependency.getId() == R.id.image) {
            Resources resources = dependency.getResources();
            final float progress = 1.f -
                    Math.abs(dependency.getTranslationY() / (dependency.getHeight() - resources.getDimension(R.dimen.collapsed_header_height)));
            final float header_height = resources.getDimension(R.dimen.collapsed_header_height);
            final float offset_y = resources.getDimension(R.dimen.collapsed_float_offset_y);
            final float child_height = child.getHeight();
            final float child_height_end = (header_height - child_height) / 2 - (header_height + offset_y);
            child.setTranslationY(header_height + resources.getDimension(R.dimen.collapsed_float_offset_y));


            // Margins
            final float collapsedMargin = resources.getDimension(R.dimen.collapsed_search_margin_right);
            final float zero = resources.getDimension(R.dimen.collapsed_float_zero);
            final float marginLeftRight = resources.getDimension(R.dimen.collapsed_search_margin_left2right);
            final int marginRight = (int) (collapsedMargin + (marginLeftRight - collapsedMargin) * progress);
            final int marginTop = (int) (child_height_end + (zero - child_height_end) * progress);
            final int marginLeft = (int) marginLeftRight;
            CoordinatorLayout.LayoutParams lp = (CoordinatorLayout.LayoutParams) child.getLayoutParams();
            lp.setMargins(marginLeft, marginTop, marginRight, 0);
            child.setLayoutParams(lp);
            return true;
        }
        return false;

    }

}

File: client/app/src/main/java/com/yszln/qiuqiu/ui/main/view/behavior/HeaderSearchBehavior.kt
package com.yszln.qiuqiu.ui.main.view.behavior

import android.content.Context
import android.util.AttributeSet
import android.view.View
import androidx.coordinatorlayout.widget.CoordinatorLayout
import com.yszln.qiuqiu.R

class HeaderSearchBehavior(context: Context?, attrs: AttributeSet?) :
    CoordinatorLayout.Behavior<View>(context, attrs) {
    override fun layoutDependsOn(
        parent: CoordinatorLayout,
        child: View,
        dependency: View
    ): Boolean {
        return dependency != null && dependency.id == R.id.image
    }

    override fun onDependentViewChanged(
        parent: CoordinatorLayout,
        child: View,
        dependency: View
    ): Boolean {
        if (dependency != null && dependency.id == R.id.image) {
            val resources = dependency.resources
            val progress = 1f -
                    Math.abs(dependency.translationY / (dependency.height - resources.getDimension(R.dimen.collapsed_header_height)))
            val header_height = resources.getDimension(R.dimen.collapsed_header_height)
            val offset_y = resources.getDimension(R.dimen.collapsed_float_offset_y)
            val child_height = child.height.toFloat()
            val child_height_end = (header_height - child_height) / 2 - (header_height + offset_y)
            child.translationY =
                header_height + resources.getDimension(R.dimen.collapsed_float_offset_y)


            // Margins
            val collapsedMargin = resources.getDimension(R.dimen.collapsed_search_margin_right)
            val zero = resources.getDimension(R.dimen.collapsed_float_zero)
            val marginLeftRight = resources.getDimension(R.dimen.collapsed_search_margin_left2right)
            val marginRight =
                (collapsedMargin + (marginLeftRight - collapsedMargin) * progress).toInt()
            val marginTop = (child_height_end + (zero - child_height_end) * progress).toInt()
            val marginLeft = marginLeftRight.toInt()
            val lp = child.layoutParams as CoordinatorLayout.LayoutParams
            lp.setMargins(marginLeft, marginTop, marginRight, 0)
            child.layoutParams = lp
            return true
        }
        return false
    }
}